### PR TITLE
nixos/buildbot: add environmentFiles option

### PR DESCRIPTION
**Pull request by aou**
This allows secrets management solutions for e.g. worker passwords

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/continuous-integration/buildbot/master.nix [hunk 1]
```diff
@@ -247,6 +247,18 @@ in {
...
       };
+
+      environmentFiles = mkOption {
+        type = types.listOf types.path;
+        default = [];
+        example = [ "/run/secrets/buildbot-master" ];
+        description = lib.mdDoc ''
+          File to load as environment file. Environment variables from this file
+          will be interpolated into the config file using envsubst with this
+          syntax: `$ENVIRONMENT` or `''${VARIABLE}`.
```
Comment (Mic92) [inline nixos/modules/services/continuous-integration/buildbot/master.nix:258]: Mhm. I got one problem with that `master.cfg` contains python code and `{FOO}` is also a valid and common python construct. Given that we have python, Is `serviceConfig.LoadCredential` not enough?

In buildbot-nix, I hook up buildbot's secret engine with systemd's `LoadCredential` like this:

```
secrets.SecretInAFile(dirname=os.environ["CREDENTIALS_DIRECTORY"])
config["secretsProviders"].append(systemd_secrets)
```

For thinks like worker passwords, I use this:

```
def read_secret_file(secret_name: str) -> str:
    directory = os.environ.get("CREDENTIALS_DIRECTORY")
    if directory is None:
        print("directory not set", file=sys.stderr)
        sys.exit(1)
    return Path(directory).joinpath(secret_name).read_text().rstrip()
```

The advantage is that environment variables are not accidentally leaked to child processes as easy through environment variables and we don't perform sneaky text replacements.

Comment (aou) [reply nixos/modules/services/continuous-integration/buildbot/master.nix:258]: ahh thank you! I thought of using buildbot's secrets but did not know you could use it in things other than steps and services. But your solution is much better. Also I should've looked in buildbot-nix.

Comment (aou) [reply nixos/modules/services/continuous-integration/buildbot/master.nix:258]: actually read_secret_file works fine, but
```
c['secretsProviders'] = [
    secrets.SecretInAFile(dirname=os.environ["CREDENTIALS_DIRECTORY"]),
]
```
is the problematic part that checks for filemode of the secrets. But, don't need to use it as I can use the read_secret method, so for now I have it commented out. Still curious how you got it working!

Comment (Mic92) [reply nixos/modules/services/continuous-integration/buildbot/master.nix:258]: I fixed the filemode stuff in buildbot and it's part of nixpkgs unstable.
It was too hard to backport, but you can get the patch from here:

https://github.com/Mic92/buildbot-nix/blob/main/nix/0001-allow-secrets-to-be-group-readable.patch

Comment (aou) [reply nixos/modules/services/continuous-integration/buildbot/master.nix:258]: Thank you!

**Pull request closed without merging**